src/dataflow/src/decode/mod.rs

```diff
@@ -377,283 +170,623 @@ fn decode_cdcv2<G: Scope<Timestamp = Timestamp>>(
-                                        "[customer-data] All upsert keys should decode to a value: {:?}",
-                                        key
-                                    ))
-                                    .into()),
-                                    Err(e) => {
-                                        Err(DecodeError::Text(format!("Error decoding key: {}", e))
-                                            .into())
-                                    }
+                    let key_cursor = &mut key.as_slice();
+                    let key = if let (Some(key_decoder), false) =
+                        (key_decoder.as_mut(), key.is_empty())
+                    {
+                        let mut key = key_decoder
+                            .next(key_cursor, None, *upstream_time_millis, false)
+                            .transpose();
+                        if let (Some(Ok(_)), false) = (&key, key_cursor.is_empty()) {
+                            // Perhaps someday we'll assign semantics to multiple keys in one message, but for now it doesn't make sense.
+                            key = Some(Err(DecodeError::Text(format!(
+                                "Unexpected bytes remaining for decoded key: {:?}",
+                                key_cursor
+                            ))
+                            .into()));
+                        }
+                        key
+                    } else {
+                        None
+                    };
+
+                    // Check whether we have a partial message from last time.
+                    // If so, we need to pre-pend it to the bytes we got from _this_ message.
+                    let value = if value_buf.is_empty() {
+                        value
+                    } else {
+                        value_buf.extend_from_slice(&*value);
+                        &value_buf
+                    };
+
+                    if value.is_empty() {
+                        session.give(DecodeResult {
+                            key,
+                            value: None,
+                            position: None,
+                        });
+                    } else {
+                        let value_cursor = &mut value.as_slice();
+                        // The intent is that the below loop runs as long as there are more bytes to decode.
+                        //
+                        // We'd like to be able to write `while !value_cursor.empty()`
+                        // here, but that runs into borrow checker issues, so we use `loop`
+                        // and break manually.
+                        loop {
+                            let old_value_cursor = *value_cursor;
+                            let value = match value_decoder.next(
+                                value_cursor,
+                                Some(n_seen + 1), // Match historical practice - files start at 1, not 0.
```

Comment:
No, but this operator only runs for "delimited" sources, so it doesn't matter.

Anyway, give me a bit, I'm trying to think of a better way to handle this line number stuff. I think we should probably just continue reporting the actual line number rather than the record number, but that might have to be a separate diff at this point.